Merge Back In Stock Notifications to WC core, pt 1: feature code

peterfabian opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR is the first WIP to initiate the merge of the [Back In Stock Notifications plugin](https://woocommerce.com/products/back-in-stock-notifications/) (BIS) to WC core.

The initial plan is to:
- Merge the code & keep it disabled for WC 9.9.
- Enable it for 5% of merchants in WC 10.0
- Enable it for all merchants in WC 10.1

#### What Does Enable Even Mean 
In order to prevent changes in the Single Product page design, the BIS, even when enabled, will have sign-ups **turned off** for out of stock products (`Allow sign-ups` unchecked). The code will be loaded, any pending notifications will be processed, settings will be accessible, etc.

Except for:
- stores that had BIS active and the setting was set to active
- new stores (approximated as stores with 0 products).

#### Data From The Past Reused

All the settings and data from the past (like customers subscribed to out of stock products) should be used correctly by the merged version. 

If merchant used BIS in the past, and disabled it at some point, then with the update to the new version of WC that includes BIS, these notifications will get processed (see decision log for discussion).

The only exception to this is the "Allow sign-ups" mentioned above. For all stores, except the ones that have BIS active & this setting set to true at the moment WC+BIS is getting installed, this setting will be set to false.

#### New Settings Location

The **BIS settings have changed their location** and will now be found under WC > Settings > Products > Customer Stock Notifications instead of WC > Settings > Stock Notifications. 

This section will be displayed once `Enable back in stock notifications for customers` under WC > Settings > Advanced > Features is checked. 

This checkbox governs the loading of the package (and thus also of all the merged BIS code) and thus acts as a kill-switch for the merged BIS code. It's linked to option `wc_feature_woocommerce_back_in_stock_notifications_enabled` that is utilized in `Packages.php` to `prep()`-are and `init()`-ialize the merged code.

This slightly differs from Brands, where the option to disable the Brands code wasn't available in the Settings UI.

#### DB initialization

BIS creates 3 db tables:
- wp_woocommerce_bis_activity
- wp_woocommerce_bis_notifications
- wp_woocommerce_bis_notificationsmeta

There are 2 options for how db tables for BIS can be created:
- during the install procedure
- during the activation of merged BIS code

This is fairly similar to how HPOS initializes its db tables.

##### DB Tables Created During Install
Same as for HPOS, BIS can add its db table schema to `WC_Install::get_schema`, if it's enabled. This way, the db tables can be created along with WC installation.

##### DB Tables Created On Activation  
Similarly to how HPOS tables are created based on the UI switch, here, it's linked to the `wc_feature_woocommerce_back_in_stock_notifications_enabled` option. When merged BIS gets activated using add/update/delete_option WordPress functions, the check for db tables is executed. If they're missing, the script tries to create them.

##### What If DB Tables Are Still Missing?
If the db tables are missing, there will be some errors in the logs, but it shouldn't end up with a fatal error.
As is the case for other WC tables, running WC > Status > Tools > `Verify base database tables` attempts to create the BIS DB tables again.

#### Activation

The activation is quite delicate and works as follows:
- the feature wp option should override the rollout settings (which is governed by BackInStockNotifications::is_enabled()).
- the feature wp option is initially without any value; in this state, the check if BIS feature is enabled falls back to the rollout setting.
- The WC > Settings > Advanced > Features UI checkbox has a default value that is set based on the rollout setting (but this is overridden by the feature wp option once it's set).
- the feature wp option is set (and later updated) when the store manager checks/unchecks the checkbox in the settings and clicks Save settings.

This way, the standalone BIS plugin can stay active for a customer with the merged BIS rollout set to false. Once the rollout is switched to yes, the standalone BIS plugin should be disabled automatically and replaced by the merged BIS instance.

There is one more thing that happens during activation: `wc_bis_daily` cron task is scheduled. 

#### Changes In The BIS Plugin

I tried to keep the changes in the BIS plugin to the minimum and merge the files as-is. I also tried to not cause any disruption to the public API of the plugin (i.e. I didn't remove any public methods, etc).

List of notable changes/things worth checking:
- I added **deprecations** to all publicly facing methods that don't make sense anymore. I might have missed some, though, as I haven't read all the code across the whole plugin. Also changed their return values to neutral values.
- I **removed some constants** that don't make sense anymore (`WC_BIS_VERSION`, `WC_BIS_SUPPORT_URL`, `WC_BIS_ABSPATH`), but that's somewhat questionable, as those are part of public interface, after all. Might need to reconsider.
- I updated the text domain to `woocommerce`
- I removed **selectSww** dependency, as that was added to solve a 1-pixel design problem. Third version of the select2 library within WC core would be too much.
- I **dropped** all the **WP Admin notices**. We might want a welcome notice, but it will probably be better to implement it using WC infrastructure.
  - this includes **loopback** tests and **Action Scheduler check** to help troubleshoot the plugin issues. These two functionalities would be better solved in the context of WC core itself.
- I **dropped** the **db update code** that copied a bunch of infrastructure from WC core, including the background processing libraries. There was also just one db update method that served to update to BIS v 1.1.0 (which was released in July 2021).
- I kept `WC_BIS()` function to make things backward compatible, but had to be more careful during initialization to not cause fatal errors (e.g. when someone attempts to activate the standalone BIS plugin while WC+BIS is active).
- I updated the time when the initialization of BIS code runs from `plugins_loaded` priority 9 to priority 10 (where it's initialized by `Packages.php`), see below for rationale.
- I updated `WC_BIS()->log()` calls to `wc_get_logger`.

##### BIS init moved from `plugins_loaded` priority 9 to priority 10

`Packages.php` has the `on_init` method hooked to `plugins_loaded` at priority 10, while the standalone BIS plugin is initialized at priority 9. It would probably be possible for me to unhook the existing action in `BackInStockNotifications::prepare` and rehook it to priority 9, then after initializing BIS rehook it back at priority 10, but it feels pretty hack-y while my guess is that it doesn't really matter too much if BIS is initialized at priority 9 or 10, so I've decided to change the BIS initialization priority instead of hacking `Packages.php`.


### PR Structure

The code has been split into 2 parts:
- the feature code that has been mostly copied from BIS plugin is part of this branch
- unit tests and e2e tests (rewritten to Playwright) have been moved to a new branch: https://github.com/woocommerce/woocommerce/pull/55836 to work around the limitation of 20k lines in one diff in GH.


### TODO:

- [x] Sort out the feature announcement
- [ ] Update since tags
- [x] Add tracking
- [ ] The plugin list looks slightly odd if there isn't any update to the BIS plugin (there's an extra border/shadow/sth)



<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

##### Build disabled zip package

1. Check out this branch of WC.
2. Open `src/Internal/BackInStockNotifications.php`, set `BackInStockNotifications::is_enabled` to return false (or it's already set to return false).
3. Build the zip file, e.g. by running `./bin/build-zip.sh` from within `plugins/woocommerce`.
4. Rename the plugin to sth like `woocommerce-9.9.0-BIS-disabled.zip`.

##### Build enabled zip package with bumped version

1. Check out this branch of WC.
2. Open `src/Internal/BackInStockNotifications.php`, set `BackInStockNotifications::is_enabled` to return true.
3. Update WC versions in woocommerce.php (plugin comment at the top) and includes/class-woocommerce.php (first property of the class) to 9.9.1. (to simulate update later)
3. Build the zip file, e.g. by running `./bin/build-zip.sh` from within `plugins/woocommerce`.
4. Rename the plugin to sth like `woocommerce-9.9.1-BIS-enabled.zip`.

##### Test Rollout Disabled to Rollout Enabled Store with BIS plugin

This test confirms that a store with standalone BIS keeps the plugin if store is assigned to the inactive cohort at first, but it switches to merged BIS once the merged code gets enabled.

1. Start with a clean environment.
2. Install the current version of WC (latest stable).
6. Install & activate BIS notifications standalone plugin.
7. Go to WC > Settings > Stock Notifications and check that the "Allow sign-ups" is checked/turned on.
8. Install the disabled zip file `woocommerce-9.9.0-BIS-disabled.zip`.
9. There should be no error visible on the front end or in the log.
10. Standalone BIS plugin should remain active.
11. The Allow sign-ups should still be checked/turned on.
12. Install the enabled zip file `woocommerce-9.9.1-BIS-enabled.zip`.
13. A notice should appear saying that BIS plugin was disabled.
14. BIS plugin should be disabled.
15. Merged BIS should be active (WC > Stock Notifications should still appear in the menu).
16. Try to activate BIS plugin.
17. You should see wp die message saying that the plugin can't be enabled.
> This plugin cannot be activated because its functionality is now included in WooCommerce core.
> Return to the Plugins page

##### Test UI option in WC > Settings > Advnaced > Features overrides Rollout Disabled 

1. Start with a clean environment.
2. Install the current version of WC (latest stable).
3. Install the disabled zip file `woocommerce-9.9.0-BIS-disabled.zip`.
4. Go to WC > Settings > Advanced > Features.
5. Check the "Enable back in stock notifications for customers" checkbox.
6. Save settings.
7. Verify that BIS functionality becomes available despite the rollout being disabled.
8. Check that WC > Settings > Products > Customer Stock Notifications is accessible.
9. Verify that BIS settings and functionality work as expected.

##### Test UI option in WC > Settings > Advnaced > Features overrides Rollout Enabled

1. Start with a clean environment.
2. Install the current version of WC (latest stable).
3. Install the enabled zip file `woocommerce-9.9.1-BIS-enabled.zip`.
4. Go to WC > Settings > Advanced > Features.
5. Uncheck the "Enable back in stock notifications for customers" checkbox.
6. Save settings.
7. Verify that BIS functionality becomes unavailable despite the rollout being enabled.
8. Check that WC > Settings > Products > Customer Stock Notifications is no longer accessible.
9. Verify that BIS settings and functionality are disabled.

##### Test Rollout Enabled State without BIS plugin---new site

This test case checks that new stores have everything enabled by default when they upgrade 

1. Start with a clean environment.
2. Install the current version of WC (latest stable).
8. Install the enabled zip file `woocommerce-9.9.1-BIS-enabled.zip`. 
9. There should be no error visible on the front end or in the log.
10. Merged BIS should be active (e.g. you should see WC > Stock Notifications menu item).
11. The Allow sign-ups should still be checked/turned on.

##### Test Rollout Enabled State without BIS plugin---existing site

This test case checks that single product page won't be changed for out of stock products for existing stores with products.

1. Start with a clean environment.
2. Install the current version of WC (latest stable).
3. Create at least one product/import the sample product CSV file.
8. Install the enabled zip file `woocommerce-9.9.1-BIS-enabled.zip`. 
9. There should be no error visible on the front end or in the log.
10. Merged BIS should be active (e.g. you should see WC > Stock Notifications menu item).
11. The Allow sign-ups should still be unchecked/turned off. 


##### Test Rollout Enabled State without BIS plugin: try to force activate BIS plugin via WP-CLI

1. Start with a clean environment.
2. Install the current version of WC (latest stable).
3. Install the enabled zip file `woocommerce-9.9.1-BIS-enabled.zip`.
4. Try to activate the standalone BIS plugin using WP-CLI:
   ```
   wp plugin activate woocommerce-back-in-stock-notifications
   ```

5. Verify that the command fails with an appropriate error message indicating the plugin cannot be activated.
6. Check that the standalone BIS plugin remains deactivated.
7. Verify that the merged BIS functionality remains active.


##### Test Rollout Enabled State with BIS plugin: verify data and settings remain from the BIS plugin (sign up for notifications and send some out)

1. Start with a clean environment.
2. Install the current version of WC (latest stable).
3. Install & activate BIS notifications standalone plugin.
4. Create a test product and set it as out of stock.
5. Sign up for back-in-stock notifications using both guest and logged-in user accounts.
6. Configure BIS settings (email templates, thresholds, etc.).
7. Install the enabled zip file `woocommerce-9.9.1-BIS-enabled.zip`.
8. Verify that:
   - All existing notification subscriptions are preserved
   - All BIS settings are maintained
   - The notification subscriptions are intact
9. Set the product back in stock and verify that notifications are sent correctly
10. Check that the notification history is maintained

##### Test Rollout Enabled State with BIS plugin: it's possible to roll back to WC version without BIS and use standalone BIS.

1. Start with a clean environment.
2. Install the current version of WC (latest stable).
3. Install & activate BIS notifications standalone plugin.
4. Create test data (products, subscriptions, settings).
5. Install the enabled zip file `woocommerce-9.9.1-BIS-enabled.zip`.
6. Verify the migration worked correctly.
7. Roll back to WC version without BIS (e.g., 9.8.x).
8. Reactivate the standalone BIS plugin.
9. Verify that:
   - All data and settings are preserved
   - The standalone plugin works as expected
   - No data loss occurred during the rollback

##### Test Rollout Enabled State with BIS plugin: check that BIS settings have moved from WC > Settings > Stock Notifications to WC > Settings > Products > Customer stock notifications

1. Start with a clean environment.
2. Install the current version of WC (latest stable).
3. Install & activate BIS notifications standalone plugin.
4. Configure various BIS settings.
5. Check WC > Settings > Stock Notifications to confirm it's there. Change a couple settings and remember it.
6. Check WC > Stock Notifications > Activity log and remember the output.
5. Install the enabled zip file `woocommerce-9.9.1-BIS-enabled.zip`.
6. Verify that:
   - The old settings location (WC > Settings > Stock Notifications) is no longer accessible
   - All settings are available in the new location (WC > Settings > Products > Customer Stock Notifications)
   - All previously configured settings are preserved in the new location
7. Verify that:
   - The old Activity tab location is no longer accessible
   - All activity data is visible in the new location (WC > Status > Customer Stock Notifications)
   - Historical activity data is preserved
7. Test saving new settings in the new location
8. Verify that the settings are properly applied


##### Check The Included BIS Works As Expected

1. Set some products out of stock.
2. Sign up to receive their back in stock notifications as logged in user and guest.
3. Set some of the products to back in stock (don't forget it needs to cross the threshold set in the BIS settings).
4. Check that the notification emails have been sent correctly (e.g. via MailPit or some other means).
5. Check that `wc_bis_daily` task is scheduled via cron.
6. Check that the REST API command works to deactivate all notifications for a product:

```
curl -X POST --location "https://example.com/wp-json/wc/v3/back-in-stock/notifications/deactivate" \
    -H "Accept: application/json" \
    -H "Content-Type: application/json" \
    -d '{
          "productId": 67
        }' \
    --basic --user ck_115:cs_204
```
7. Ensure the unit tests included in the PR pass.
8. Ensure the e2e tests included in the PR pass.




--




<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [x] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Added Back In Stock Notifications functionality to WC core.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

peterfabian committed:
Initial BIS code merge.

peterfabian committed:
Added main plugin file (verbatim for now).

peterfabian committed:
Hooked the singleton to WC class.

Kept the WC_BIS() global function for bw compatibility, but some things might not make sense anymore (lie plugin version), so this is of questionable benefit.

peterfabian committed:
Deprecate compatibility version comparison methods.

peterfabian committed:
We probably won't be storing BIS db version anymore.

peterfabian committed:
Handles plugin version (and path prefix) in script registering calls.

peterfabian committed:
Corrected path to main file.

peterfabian committed:
Assets path mapped.

peterfabian committed:
Make the naming of enabled functions more explicit.

peterfabian committed:
Init db test.

peterfabian committed:
Fix the infinite loop when checking for enabled BIS.

peterfabian committed:
Let core know about the package.

peterfabian committed:
Cleaned up BIS Install class.

peterfabian committed:
Compatibility classes not needed anymore.

peterfabian committed:
Update notices location and inclusion path.

peterfabian committed:
Future db updates will be handled within core.

peterfabian committed:
Remove the background updater refs.

peterfabian committed:
Added neutral return values to deprecated methods.

peterfabian committed:
Removed testing filters.

peterfabian committed:
WC_BIS_ABSPATH clean up.

peterfabian committed:
CLI db update no longer needed, hadnled by core.

peterfabian committed:
get_plugin_path is now deprecated.

peterfabian committed:
Stylesheet path & name fixed.

peterfabian committed:
Prevent activation errors.

peterfabian committed:
DB update methods no longer needed.

peterfabian committed:
Update db notices will be handled by core.

peterfabian committed:
Further improvements to the activation.

peterfabian committed:
Ensure db create routine works in context of WC core.

peterfabian committed:
Check BIS db tables on each request and create tables if missing.

peterfabian committed:
Handle activation when updating WC core to WC+BIS.

peterfabian committed:
Further defatilze activation. Sprinkled legacy proxy on top.

peterfabian committed:
Translation domain updated.

peterfabian committed:
Improved daily task handling.

peterfabian committed:
Removed loopback and queue notices.

peterfabian committed:
Removed leftover notice code.

peterfabian committed:
Deprecate and stop using wc_bis_get_formatted_screen_id.

peterfabian committed:
Settings revamp.

peterfabian committed:
Removed selectsw library.

peterfabian committed:
Updated db create routine to be called on option update.

peterfabian committed:
Manual mapping of feature option doesn't seem necessary.

peterfabian committed:
Disable signups for stores that didn't have BIS active.

peterfabian committed:
Apply changes from BIS PR 353

peterfabian committed:
Applied changes from BIS PR 341

peterfabian committed:
Added unit tests.

peterfabian committed:
Applied changes from BIS PR 349

peterfabian committed:
Applied changes from BIS PR 363

peterfabian committed:
Applied changes from BIS PR 368

peterfabian committed:
Added BIS framework to core unit tests.

peterfabian committed:
Fixed email include paths after BIS PR 363.

peterfabian committed:
Change sign-ups setting only for enabled WC+BIS.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

peterfabian commented:
Some jobs failed because the diff it too large, not sure what to do about it.

> Error: Command failed: gh pr diff 55787 --name-only
could not find pull request diff: HTTP 406: Sorry, the diff exceeded the maximum number of lines (20000) (https://api.github.com/repos/woocommerce/woocommerce/pulls/55787)
PullRequest.diff too_large

peterfabian committed:
Removed welcome notice.

peterfabian committed:
Renamed class back to WC_Back_In_Stock to maintain bw compatibility.

peterfabian commented:
I tried to decrease the number of lines below 20k, but apparently it's not only about changed lines (+/-), but also includes context lines, so this needs further cutting.

peterfabian committed:
Make BIS disabled for rollout.

peterfabian committed:
Revert "Added BIS framework to core unit tests."

This reverts commit 723c104b2b0ac82a83338311c8ce477e1eb76278.

peterfabian committed:
Revert "Added unit tests."

This reverts commit e3c03984e309d93c998cbf3cd9193fe0a6674ef7.

peterfabian committed:
Fixed unit test for Products settings.

peterfabian committed:
Avoid superfluous label in settings and thus fix e2e test.

peterfabian committed:
Removed Status page that doesn't do anything.

peterfabian committed:
Fixed e2e test to reflect settings changes.

peterfabian committed:
Fixed another e2e test to reflect settings change.

peterfabian committed:
Bumped template versions.

coreymckrill commented on plugins/woocommerce/src/Internal/BackInStockNotifications.php:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	/**
+	 * Deactivate signups for BIS to prevent changing the single product screen.
```
I suspect this description will not age well. What does "changing the single product screen" mean in this case?

peterfabian replied:
If the sign-ups are allowed, there's a change in the single product screen: a new element for buyers to get notified once the product is back in stock.

<img width="1724" alt="Screenshot 2025-02-18 at 10 20 14" src="https://github.com/user-attachments/assets/85b7cf8f-2c2b-4b7e-96dc-6b2de2b0b096" />

We wanted to avoid changing this screen for merchants just by updating Woo core, as it might be customized.

coreymckrill replied:
Ah, gotcha. I think it would be beneficial to make this doc block more explicit so that in a couple of years we can still understand what this was talking about.

peterfabian replied:
Thanks, that's good feedback. After the design review, it seems like this might be changing, so we'll see where we end up with this.

coreymckrill reviewed (commented):
> Testing Disabled State

I followed the instructions, and built the zip of WC with the `is_enabled` method returning false (it was already set to do this when I checked out the branch). However, when installing the new version of WC, it appears that BIS was automatically disabled, and shows the message `This plugin can no longer be activated because its functionality is now included in WooCommerce. It is recommended to delete it.` But the BIS functionality in WC is also turned off, since `is_enabled` returns false.

A couple of other initial thoughts:

* Why is the bulk of the BIS code getting added to `includes` instead of `src`? Is it because it's not namespaced? Ideally we wouldn't be adding new code in `includes`, but I suppose it might be difficult to adapt everything to namespaces if the autoloader for the `src` folder requires their usage.

peterfabian committed:
Removed forgotten SW Select dependency.

peterfabian commented:
> However, when installing the new version of WC, it appears that BIS was automatically disabled, and shows the message This plugin can no longer be activated because its functionality is now included in WooCommerce. It is recommended to delete it. But the BIS functionality in WC is also turned off, since is_enabled returns false.

Good catch. I think this happens because we fairly recently decided to switch the feature flag option `wc_feature_woocommerce_back_in_stock_notifications_enabled` to a setting (under the WC > Settings > Products > Inventory) and that means it gets a default value `yes`, which makes `Automattic\WooCommerce\Packages` class [think it's enabled](https://github.com/woocommerce/woocommerce/blob/mic/bis-merge-take-2/plugins/woocommerce/src/Packages.php#L135-L138) irrespective of the `is_enabled()` method. I need to think about how to handle this case correctly. Changing the Packages isn't easy as Brands code is already using it, too. Will need to think about the best way to handle this.

> Why is the bulk of the BIS code getting added to includes instead of src? Is it because it's not namespaced? Ideally we wouldn't be adding new code in includes, but I suppose it might be difficult to adapt everything to namespaces if the autoloader for the src folder requires their usage.

I've reused the same approach that was used to merge Brands plugin. Previous discussions on this topic can be found in the slack channel: p1721257169476979-slack-C07DH5W04L8 and comments below, but the tl;dr is that the code fits best into the `includes` folder and its subdirectories, as it doesn't follow PSR-4 and doesn't have namespaces. Some more discussion also wrt to autoloading over here: p1725283129091499-slack-C07DH5W04L8. Hope that makes sense.

coreymckrill commented on plugins/woocommerce/includes/bis/class-wc-bis-sync.php:
```diff
@@ -0,0 +1,274 @@
+		}
+
+		if ( wc_bis_debug_enabled() ) {
+			WC_BIS()->log( 'Product IDs: ' . print_r( $this->queue, true ), 'info', 'wc_bis_sync_logs' );
```
It would be nice to just use `wc_get_logger()` directly here, and also clean this up to use [logging best practices](https://developer.woocommerce.com/docs/logging-in-woocommerce/#3-logging-best-practices) (i.e. don't output the array into the log message).

peterfabian replied:
Makes sense, thanks. Updated in [6508003](https://github.com/woocommerce/woocommerce/pull/55787/commits/6508003c3d0f1fc10032af4249040b53cfeeb0e0).

coreymckrill commented on plugins/woocommerce/includes/bis/class-wc-bis-sync-tasks.php:
```diff
@@ -0,0 +1,379 @@
+		}
+
+		if ( wc_bis_debug_enabled() ) {
+			WC_BIS()->log( sprintf( 'Sending %d notifications for products %s ', $notifications_count, implode( ',', $product_ids ) ), 'info', 'wc_bis_sync_logs' );
```
It would be nice to just use `wc_get_logger()` directly here, and also clean this up to use [logging best practices](https://developer.woocommerce.com/docs/logging-in-woocommerce/#3-logging-best-practices).

peterfabian replied:
You're right, updated in [6508003](https://github.com/woocommerce/woocommerce/pull/55787/commits/6508003c3d0f1fc10032af4249040b53cfeeb0e0).

coreymckrill commented on plugins/woocommerce/includes/admin/class-wc-bis-admin.php:
```diff
@@ -0,0 +1,513 @@
+		$suffix = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ? '' : '.min';
+		$version   = Constants::get_constant( 'WC_VERSION' );
+
+		wp_register_style( 'wc-bis-admin', WC()->plugin_url() . '/assets/css/bis-admin.css', array( 'sw-admin-css-select' ), $version );
+		wp_style_add_data( 'wc-bis-admin', 'rtl', 'replace' );
+
+		wp_register_script( 'wc-bis-writepanel', WC()->plugin_url() . '/assets/js/admin/wc-bis-admin' . $suffix . '.js', array( 'jquery', 'jquery-ui-datepicker', 'wp-util', 'sw-admin-select-init', 'wc-backbone-modal' ), $version );
```
I'm seeing some errors in Query Monitor related to missing dependencies:

<img width="1059" alt="image" src="https://github.com/user-attachments/assets/872524fc-e554-4a31-9630-ae250c7c7b84" />

<img width="1130" alt="image" src="https://github.com/user-attachments/assets/94bed5ca-d347-427e-bdf1-efdd2c3e2fb4" />

peterfabian replied:
Well spotted! I fixed this yesterday but forgot to push. Fixed in [067fdda](https://github.com/woocommerce/woocommerce/pull/55787/commits/067fdda85649fc099a9cd50e6b6392842c6a8537).

peterfabian committed:
Roll back the previous change in WC Settings > Products > Inventory.

peterfabian committed:
Added BIS setting to Advanced > Features.

peterfabian committed:
Move WC > Stock Notifications > Activity to WC > Status.

peterfabian committed:
Updated the activation code to correctly handle all cases.

peterfabian committed:
Hooked into Features > Advanced without changing the controller.

peterfabian committed:
Fixed the check on when the merged BIS code is enabled.

peterfabian committed:
Revert "Hooked into Features > Advanced without changing the controller."

This reverts commit a29076172783c3f09f459317d99d777dd55114a7.

peterfabian committed:
Rollout starts with BIS disabled.

peterfabian committed:
Fixed allow sign-ups seting for new installations.

peterfabian committed:
Make sure the feature is only enabled for active rollout customers.

peterfabian committed:
Switched logging to wc core's logger.

peterfabian committed:
Updated unit tests to reflect the change in Products > Inventory settings.

peterfabian committed:
Revert "Fixed e2e test to reflect settings changes."

This reverts commit 22353c4db5e2e76a7f28d5d555fce91b2dbb8754.

peterfabian committed:
Revert "Fixed another e2e test to reflect settings change."

This reverts commit bb7f317bca73e4a688ca645d1a0a3ebcf023d422.

peterfabian committed:
Fixed the init procedure via Settings UI.
When the feature was disabled, the hooks to set up the infrastructure were not hooked, so I added a way to run some code even when BIS is disabled.

jimjasson commented:
@peterfabian 

I started testing your PR and here's what I found so far: 

(1) Existing site with Back In Stock Notifications enabled 

I tested this scenario extensively and I confirmed that:
- The Back In Stock Notifications plugin remained active, while core's module was enabled. 
- When I enabled the core's module via the UI, Back In Stock Notifications was automatically disabled and a notice showed up. 
- When I tried to enable Back In Stock Notifications via the Plugins page, it wasn't possible. 
- When I tried to enable Back In Stock Notifications via CLI: `wp plugin activate woocommerce-back-in-stock-notifications` and `wp plugin activate woocommerce-back-in-stock-notifications --skip-plugins=woocommerce`, the plugin remained deactivated. 
- All the settings I had in Back In Stock Notifications were transferred to the Inventory tab and their values were honored. 
- When I changed the settings, disabled the core module and re-enabled the plugin, all changed settings were transferred back to the plugin settings.

The only thing I found a bit confusing as an existing user is that the settings changed location without any notice -- perhaps the notice that says that the Back In Stock Notifications has been disabled, should also mention the settings changing location?

(2) New sites

In a new site, I installed this WooCommerce branch and enabled Back In Stock Notifications via the UI, but the DB options weren't created.

peterfabian commented:
> (2) New sites

> In a new site, I installed this WooCommerce branch and enabled Back In Stock Notifications via the UI, but the DB options weren't created.

Thank a lot for testing, @jimjasson !

I believe I fixed this problem in [946eb26](https://github.com/woocommerce/woocommerce/pull/55787/commits/946eb2633ea2c274cb126858c92c3b3a19498055).

jimjasson commented:
Thanks @peterfabian! I confirmed that after the fix, the DB tables and the cron job are added to the site correctly.

peterfabian committed:
Merge branch 'trunk' into mic/bis-merge-take-2

peterfabian committed:
Removed superfluous checks.

peterfabian committed:
Moved settings to BISN class.

WC core adds all options with its default value to the database. However, this doesn't work well for merged BIS code where the db option acts as an override for the rollout assignment. This trick relies on the default behaviour of WP settings to use the default value (which can change in code) until the underlying option is set.

peterfabian committed:
Fixed leftover after renaming WC_BIS_Notifications.

peterfabian committed:
Fixed feature option handling during WC updates.

peterfabian committed:
Ensure scheduled job gets created during update.

coreymckrill commented on plugins/woocommerce/includes/admin/class-wc-bis-admin.php:
```diff
@@ -0,0 +1,513 @@
+			wp_localize_script( 'wc-bis-writepanel', 'wc_bis_admin_params', $params );
+		}
+
+		wp_enqueue_style( 'wc-bis-admin' );
```
Does this stylesheet need to get enqueued on every admin page?

coreymckrill commented on plugins/woocommerce/includes/admin/class-wc-bis-admin.php:
```diff
@@ -0,0 +1,513 @@
+		 * Enqueue specific styles & scripts.
+		 */
+		if ( WC_BIS()->is_current_screen( array( 'woocommerce_page_wc-settings' ) ) ) {
+			wp_enqueue_script( 'wc-bis-writepanel' );
```
Does this script need to get enqueued on every WC Settings page?

jimjasson commented:
Thank you for the latest changes @peterfabian! 

I re-tested the upgrade + activation flow and confirmed that:
- When I have a new site with WooCommerce with the Back In Stock Notifications disabled and I upgrade WooCommerce to a version with Back In Stock Notifications enabled, then both the DB tables and the Cron job are created. Moreover, the "Allow sign-ups" option is enabled. 
- When I have an existing site with WooCommerce with the Back In Stock Notifications disabled and I upgrade WooCommerce to a version with Back In Stock Notifications enabled, then both the DB tables and the Cron job are created. Moreover, the "Allow sign-ups" option is disabled.

In both scenarios, the site didn't have the Back In Stock Notifications plugin.

jimjasson commented:
Looking at the code changes @peterfabian, I have the following concerns (from minor to more important);

(1) The `@since` and `@version` numbers in all files are still pointing to the Back In Stock Notifications plugin versions. These should be updated to either `x.x.x` or to the WooCommerce version you are targeting. 

(2) The Back In Stock Notifications file names don't follow core's naming conventions. Compare `class-wc-admin-brands` with `class-wc-bis-admin-ajax` -- I think it should become `class-wc-admin-bis-ajax`. Similar naming changes were made when merging Brands into core. 

(3) The `prepare_always` and `is_really_enabled` functions 

After our sync discussion and re-reading the code, I understand that with the Back In Stock Notifications package there is the need to run some code even when the package is disabled. For example, you want to register the **WooCommerce > Settings > Advanced > Features > Back In Stock Notifications** option, without loading the rest of the Back In Stock Notifications code. 

Therefore, `prepare`, which currently runs only for enabled packages isn't enough: https://github.com/woocommerce/woocommerce/blob/2b51d8248c746091057d1cefe5d7df6d8ab2ddc4/plugins/woocommerce/src/Packages.php#L172 and you would like a function similar to `prepare` that runs for all merged packages. 

Given that `prepare` was first added to the Brands module, my proposal here is to make it run for all merged packages (only Brands and Back In Stock Notifications are merged packages for now), even when they are not enabled. Then, inside each `prepare` distinguish which code will run for when the package is enabled and which code will run always regardless of the status of the package. 

As for `is_really_enabled`, I understand that the need for such a function stems for the fact that `is_enabled` essentially check if there are any restrictions tied to our gradual rollout plan. Looking back at Brands I think `is_enabled` was a poor naming choice and this should change in both packages so that:
- `is_enabled` returns `Packages::is_package_enabled( self::$PACKAGE_NAME );` and;
- a new function should be created to check if the package should be enabled/disabled based on the rollout plan. 

pinging @xristos3490 for his thoughts on this as well.

peterfabian commented:
Thanks for the comments, @jimjasson !

> (1) The @since and @version numbers in all files are still pointing to the Back In Stock Notifications plugin versions. These should be updated to either x.x.x or to the WooCommerce version you are targeting.

Agreed, it's still in the TODO list in the PR description:

> TODO:
> - Sort out the feature announcement
> - Update since tags <---- HERE
> - Add tracking
> - The plugin list looks slightly odd if there isn't any update to the BIS plugin (there's an extra border/shadow/sth)

> (2) The Back In Stock Notifications file names don't follow core's naming conventions. Compare class-wc-admin-brands with class-wc-bis-admin-ajax -- I think it should become class-wc-admin-bis-ajax. Similar naming changes were made when merging Brands into core.

I'm a bit split here. Were you not afraid of breaking backward compatibility? Also, to me, it feels more important to retain the convention where file name maps to class name (and vice versa) than to rename wc-bis-admin... to wc-admin-bis...

> (3) The prepare_always and is_really_enabled functions

Let's discuss with others, too. I agree there are multiple options on how to handle these abstractions.

I agree with changing the semantics of `is_enabled` and adding another method to handle the rollout enabled/disabled states.

As for `prepare_always`, I don't have a strong opinion.

peterfabian commented:
Updated the testing instructions to (hopefully) cover activation scenarios more clearly.

peterfabian commented:
Thanks for your input, @coreymckrill !

> > Database
>
> As above, thinking about the longer-term past the initial rollout, would it make sense to consolidate the database schema into WC_Install so that future us can see at a glance all the extra tables that WC generates, rather than having to go hunting for them in multiple files?

I can definitely do that. I tried to follow the example of HPOS and product lookup tables and only added the tables to the schema if the feature is enabled, but perhaps we want to approach this differently? Given that this will be happening more in the future, should we create some kind of guideline about how we want to approach this? (e.g. also discussed in p7bje6-6UG-p2).

> I left a couple of notes about scripts and stylesheets that probably don't need to be loaded everywhere.

Thanks, I fixed those in

peterfabian committed:
Removed CSS for already removed admin notices.

peterfabian committed:
Don't enqueue styles and scripts everywhere in admin.

peterfabian committed:
Fixed the Activity page and related SwwSelect issues.

peterfabian committed:
Removed extra whitespace.

peterfabian committed:
Added the basic Tracker option for BIS.

peterfabian committed:
Fixed the forgotten rollout enabled switch.

peterfabian committed:
Renamed the confusingly named is_enabled method (also for Brands).

peterfabian commented:
> As for is_really_enabled, I understand that the need for such a function stems for the fact that is_enabled essentially check if there are any restrictions tied to our gradual rollout plan. Looking back at Brands I think is_enabled was a poor naming choice and this should change in both packages so that:
>
> - is_enabled returns Packages::is_package_enabled( self::$PACKAGE_NAME ); and;
> - a new function should be created to check if the package should be enabled/disabled based on the rollout plan.

Implemented in [87dacbb](https://github.com/woocommerce/woocommerce/pull/55787/commits/87dacbb42587becf75bfb0c15b36dea9bfe2ac12).

peterfabian committed:
Merge branch 'trunk' into mic/bis-merge-take-2

xristos3490 commented on plugins/woocommerce/includes/bis/compatibility/core/class-wc-bis-core-compatibility.php:
```diff
@@ -0,0 +1,191 @@
+	 * @return boolean
+	 */
+	public static function wc_current_theme_is_fse_theme() {
+		return function_exists( 'wc_current_theme_is_fse_theme' ) ? wc_current_theme_is_fse_theme() : false;
```
I think this should also be deprecated. After all, it's another Core global function. It's also used [in the Templates controller](https://github.com/woocommerce/woocommerce/pull/55787/files#diff-c32c0e4eb9ef79a55cf6a9c277ab869b5434723090569076431c21bc6fb5511fR74) when we load the frontend assets.

peterfabian replied:
Good point, updated in [bb8ca00](https://github.com/woocommerce/woocommerce/pull/55787/commits/bb8ca005f7e36d67819e78926bebd24607423996).

xristos3490 commented on plugins/woocommerce/includes/admin/settings/class-wc-bis-admin-settings.php:
```diff
@@ -0,0 +1,348 @@
+					'type'  => 'title',
+				);
+
+				if ( ! BackInStockNotifications::is_enabled() ) {
```
I don't know how we can end up here. Considering this controller won’t function if the feature is disabled, right?

peterfabian replied:
You're right, that's an atavism from when the settings were handled differently. Removed in [3d439dd](https://github.com/woocommerce/woocommerce/pull/55787/commits/3d439ddc32efc74722c6f17fe87d1c2311c6467b).

xristos3490 reviewed (commented):
Fantastic work so far @peterfabian! 👏  Before looking into the Packages controller, I found a few issues that need addressing. 

The first issue involves the following actions:
- Install the BIS plugin.
- Use this branch.
- With the BIS plugin activated, try to enable the feature from the "Advanced > Features" checkbox.
- This results in an error because the proxy still accesses the plugin version of `WC_BIS_Install`, where the schema method is private. Haven't looked for solutions, but the most obvious one I guess would be to use a reflection here.

The error:
```
[11-Mar-2025 16:56:01 UTC] PHP Fatal error:  Uncaught TypeError: call_user_func_array(): Argument #1 ($callback) must be a valid callback, cannot access private method WC_BIS_Install::get_schema() in /Users/chris/Repos/woocommerce/plugins/woocommerce/src/Proxies/LegacyProxy.php:98
Stack trace:
#0 /Users/chris/Repos/woocommerce/plugins/woocommerce/src/Proxies/LegacyProxy.php(98): call_user_func_array('WC_BIS_Install:...', Array)
#1 /Users/chris/Repos/woocommerce/plugins/woocommerce/src/Internal/BackInStockNotifications.php(327): Automattic\WooCommerce\Proxies\LegacyProxy->call_static('WC_BIS_Install', 'get_schema')
#2 /Users/chris/Repos/woocommerce/plugins/woocommerce/src/Internal/BackInStockNotifications.php(374): Automattic\WooCommerce\Internal\BackInStockNotifications::bis_tables_exist()
#3 /Users/chris/Repos/woocommerce/plugins/woocommerce/src/Internal/BackInStockNotifications.php(451): Automattic\WooCommerce\Internal\BackInStockNotifications::maybe_create_database_tables()
#4 /Users/chris/Repos/woocommerce/plugins/woocommerce/src/Internal/BackInStockNotifications.php(393): Automattic\WooCommerce\Internal\BackInStockNotifications::maybe_update_bis_infrastructure(NULL, 'yes', 'wc_feature_wooc...')
#5 /Users/chris/Sites/woo/wp-includes/class-wp-hook.php(324): Automattic\WooCommerce\Internal\BackInStockNotifications::handle_add_option('wc_feature_wooc...', 'yes')
#6 /Users/chris/Sites/woo/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters('', Array)
#7 /Users/chris/Sites/woo/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#8 /Users/chris/Sites/woo/wp-includes/option.php(1168): do_action('add_option_wc_f...', 'wc_feature_wooc...', 'yes')
#9 /Users/chris/Sites/woo/wp-includes/option.php(921): add_option('wc_feature_wooc...', 'yes', '', 'on')
#10 /Users/chris/Repos/woocommerce/plugins/woocommerce/includes/admin/class-wc-admin-settings.php(993): update_option('wc_feature_wooc...', 'yes', 'yes')
#11 /Users/chris/Repos/woocommerce/plugins/woocommerce/includes/admin/settings/class-wc-settings-page.php(532): WC_Admin_Settings::save_fields(Array)
#12 /Users/chris/Repos/woocommerce/plugins/woocommerce/includes/admin/settings/class-wc-settings-advanced.php(542): WC_Settings_Page->save_settings_for_current_section()
#13 /Users/chris/Sites/woo/wp-includes/class-wp-hook.php(324): WC_Settings_Advanced->save('')
#14 /Users/chris/Sites/woo/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters('', Array)
#15 /Users/chris/Sites/woo/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#16 /Users/chris/Repos/woocommerce/plugins/woocommerce/includes/admin/class-wc-admin-settings.php(89): do_action('woocommerce_set...')
#17 /Users/chris/Repos/woocommerce/plugins/woocommerce/includes/admin/class-wc-admin-menus.php(186): WC_Admin_Settings::save()
#18 /Users/chris/Sites/woo/wp-includes/class-wp-hook.php(324): WC_Admin_Menus->save_settings('')
#19 /Users/chris/Sites/woo/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array)
#20 /Users/chris/Sites/woo/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#21 /Users/chris/Sites/woo/wp-settings.php(726): do_action('wp_loaded')
#22 /Users/chris/Sites/woo/wp-config.php(102): require_once('/Users/chris/Si...')
#23 /Users/chris/Sites/woo/wp-load.php(50): require_once('/Users/chris/Si...')
#24 /Users/chris/Sites/woo/wp-admin/admin.php(34): require_once('/Users/chris/Si...')
#25 /Users/chris/.composer/vendor/laravel/valet/server.php(110): require('/Users/chris/Si...')
#26 {main}
  thrown in /Users/chris/Repos/woocommerce/plugins/woocommerce/src/Proxies/LegacyProxy.php on line 98

```

The second one is related to the admin menu. See the double highlighted items under "WooCommerce":
![Screenshot 2025-03-11 at 7 00 06 PM](https://github.com/user-attachments/assets/a132d042-563e-40a0-abe7-812911e637d0)

This is caused by a few tweaks we made to modify the highlighted item in this jQuery script: https://github.com/woocommerce/woocommerce-back-in-stock-notifications/blob/536b6d72edac3197f3ebd9ec14581caa6e624d5e/resources/js/admin/wc-bis-admin.js#L10-L16 To solve this properly I think we need to:
- Remove the script above.
- Fix the "Notifications" (second tab) submenu to highlight the correct parent item with some fixing in some globals like the `global $parent_item`.

peterfabian committed:
Handled edge case when BIS plugin is still loaded but merged BIS needs db schema.

xristos3490 commented on plugins/woocommerce/includes/bis/class-wc-bis-templates.php:
```diff
@@ -0,0 +1,136 @@
+		// Styles
+		wp_register_style( 'wc-bis-css', WC()->plugin_url() . '/assets/css/frontend/bis.css', false, $version, 'all' );
+		wp_style_add_data( 'wc-bis-css', 'rtl', 'replace' );
+		wp_enqueue_style( 'wc-bis-css' );
```
Something else to consider is that we load a single CSS for each frontend view. If a block theme is active, then we also load another one. Honestly, regarding this merge, I think it would be better to: 
- Repurpose the `dequeue_product_scripts` function to also handle styles, 
- Merge the block theme with the main CSS (possibly by prefixing rules with the `woocommerce-uses-block-theme` class).

WDYT? @peterfabian

peterfabian replied:
Sounds like a good suggestion. Something like this perhaps? [458101e](https://github.com/woocommerce/woocommerce/pull/55787/commits/458101e0f5baa15d681bbbe15390e010d0629db6)

xristos3490 replied:
Looking good! Left a comment here https://github.com/woocommerce/woocommerce/pull/55787/commits/458101e0f5baa15d681bbbe15390e010d0629db6#r1995311931

peterfabian committed:
Allow early plugin disabling and deferred notice about disabling plugin to support redirects.

peterfabian committed:
Prevented incorrect submenu highlighting after moving Activity page.

peterfabian committed:
Deprecated duplicate wc_current_theme_is_fse_theme.

peterfabian committed:
Removed obsolete settings checks added during the merge.

peterfabian committed:
Fixed settings missing JS.

peterfabian committed:
Merged frontend CSS styles. Dequeued when not needed.

peterfabian committed:
Removed deprecated method call.

peterfabian committed:
Fixed CSS loading path.

peterfabian committed:
Pleased the JS linter.

peterfabian commented:
> The linter seems happy.

Heh, seems like it was showing up as ✅ only because it had a fatal during the run. Will fix the complaints now that it actually runs.

peterfabian committed:
Check the nonce for search correctly, plus linter update.

peterfabian committed:
Formatted strict_types correctly.

peterfabian committed:
Lint fixed, removed notices code that was no longer relevant.

peterfabian committed:
Lint fixes in class-wc-bis-admin-dashboard-page.php

peterfabian committed:
Applied PHPCBF.

peterfabian committed:
Fixed link to settings that moved elsewhere.

peterfabian committed:
Switch WC_BIS_Admin_Notices::add_notice to wp_admin_notice and WC_Admin_Notices::add_notice.

xristos3490 commented on plugins/woocommerce/client/legacy/css/bis.scss:
```diff
@@ -0,0 +1,68 @@
+}
+
+// Single product fix.
+div.product .wc_bis_product_form {
```
To avoid any visual regressions I would suggest prefixing here with block theme's frontend wrapper. So:
```suggestion
.woocommerce-uses-block-theme div.product .wc_bis_product_form {
```

xristos3490 replied:
Same for the second rule of the old `blocktheme.css`

peterfabian replied:
Makes sense, done in 5a6a6000402616fff3064b854340ea561462dfa2.

xristos3490 commented on plugins/woocommerce/includes/bis/class-wc-bis-ajax.php:
```diff
@@ -0,0 +1,28 @@
+ * @class    WC_BIS_Ajax
+ * @version  1.0.0
+ */
+class WC_BIS_Ajax {
```
I believe this is noop and can be removed.

peterfabian replied:
Correct, removed.

xristos3490 commented on plugins/woocommerce/src/Packages.php:
```diff
@@ -170,6 +177,13 @@ public static function is_package_enabled( $package ) {
+		// Prepare all merged packages for initialization even if they are disabled.
+		foreach ( self::$merged_packages as $package_name => $package_class ) {
+			if ( method_exists( $package_class, 'prepare_always' ) ) {
+				call_user_func( array( $package_class, 'prepare_always' ) );
```
While this can be very opinionated stuff, it seems to me that a more appropriate name would be `setup` here.

peterfabian replied:
I like the suggestion, renamed in [acbcded](https://github.com/woocommerce/woocommerce/pull/55787/commits/acbcdedcf5e17054fdb201a08e0bb5f1db6221be).

xristos3490 commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -1188,6 +1190,24 @@ public function mailer() {
+	 *
+	 * @return WC_Back_In_Stock|WC_BIS_Noop
+	 */
+	public function bis_notifications() {
```
Where is this used? 🤔

peterfabian replied:
Nowhere at the moment. I was thinking whether to include this so that devs would get the BIS using the `WC()-> bis_notifications()` method instead of `WC_BIS()`. Similar to `queue`, `shipping`, `mailer`. What do you see as preferable?

xristos3490 replied:
Interesting! My two cents are that the `WC_BIS()` as a standalone namespace is preferable for now. I feel we should only integrate well-abstracted utilities/traits into that global function or have a strong reason for doing so?

peterfabian replied:
You're probably right that it's unnecessary. Removed in [69ebea3](https://github.com/woocommerce/woocommerce/pull/55787/commits/69ebea3edcfdc59b467098db7144b94df04e141e).

xristos3490 commented on plugins/woocommerce/src/Internal/BackInStockNotifications.php:
```diff
@@ -0,0 +1,516 @@
+		// Check for CLI activation via WP-CLI
+		if ( defined('WP_CLI') && WP_CLI ) {
+			global $argv;
+			if ( is_array( $argv ) && in_array( 'plugin', $argv ) && in_array( 'activate', $argv ) && in_array( $bis_plugin_name, $argv ) ) {
```
If I want to activate bis via CLI, I'm gonna probably write this:
```
wp plugin activate woocommerce-back-in-stock-notifications
```
Which doesn't include the PHP filename thus it falls to: "Cannot redeclare WC_BIS() etc".

peterfabian replied:
Well spotted, I haven't tested this so far, so thanks for noticing!

peterfabian replied:
Updated.

xristos3490 reviewed (commented):
Shared some more findings/questions while I am wrapping up manual testing.

peterfabian committed:
Removed WC_BIS_Ajax that's no longer needed.

peterfabian committed:
First batch of PHPCS fixes.

peterfabian committed:
Second batch of PHPCS fixes.

peterfabian committed:
Merge branch 'trunk' into mic/bis-merge-take-2

peterfabian committed:
Third batch of PHPCS fixes.

peterfabian committed:
Fourth batch of PHPCS changes.

peterfabian committed:
Fixed the product delete hook arg count.

peterfabian committed:
Retain the priority of the hook.

peterfabian committed:
Made the block styles only apply in case woocommerce-uses-block-theme class is used.

peterfabian committed:
Renamed prepare_always to setup to make the name better.

peterfabian committed:
Removed the WC object's reference to BIS instance.

peterfabian committed:
Fixed dbDelta idempotency.

peterfabian committed:
Updated since and version tags.

peterfabian committed:
Pleasing dbDelta some more.

peterfabian committed:
Merge branch 'trunk' into mic/bis-merge-take-2

peterfabian commented:
As it was decided to reimplement this feature in a different way (integrate more deeply with core and not include everything in the legacy `includes` folder), this PR is no longer relevant.
